From a3ba3c0b9143c99376431e8d890b6e27fb889add Mon Sep 17 00:00:00 2001
From: Patrick Combes <patrick.combes@intel.com>
Date: Mon, 6 May 2013 17:12:31 +0200
Subject: [PORT FROM MAIN][BT] Add BCM43241 Bluetooth chip autodetection

BZ: 104623

Port patch device-common-libbt.patch provided by Broadcom
(see CSP case 648376).

This patch checks the local_chip_name field to detect BCM43241 and then
select the right firmware based on lmp_subversion.

Note that HCD file is still labelled as BCM43241B3 for B3/B4 revision.

Change-Id: Ied7bfafcc339129889bb2c7734e5be7847fffd40
Orig-Change-Id: I02b2127a7a12a010b77b19cdcb5b3fcd7373edc2
Feature: BT
Category: Integration
Purpose: ABSP-Adapter
Origin: 3rdVendors
Upstream-candidate: no
Signed-off-by: Patrick Combes <patrick.combes@intel.com>
Reviewed-on: http://android.intel.com:8080/106412
Reviewed-by: cactus <cactus@intel.com>
Reviewed-by: Boniface, XavierX <xavierx.boniface@intel.com>
Reviewed-by: Bondier, CedricX <cedricx.bondier@intel.com>
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Reviewed-by: Gros, Julien <julien.gros@intel.com>
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
---
 libbt/src/hardware.c |   21 ++++++++++++++++-----
 1 files changed, 16 insertions(+), 5 deletions(-)

diff --git a/libbt/src/hardware.c b/libbt/src/hardware.c
index 3ce3e55..98b4eba 100644
--- a/libbt/src/hardware.c
+++ b/libbt/src/hardware.c
@@ -809,9 +809,10 @@ void hw_config_cback(void *p_mem)
                 hw_cfg_cb.local_chip_name[LOCAL_NAME_BUFFER_LEN-1] = 0;
 
                 /* Additional check for revision if chip is BCM4335 */
-                if (strstr(hw_cfg_cb.local_chip_name, "BCM4335") != NULL)
+                if ( (strstr(hw_cfg_cb.local_chip_name, "BCM4335") != NULL) ||
+                    (strstr(hw_cfg_cb.local_chip_name, "BCM43241") != NULL) )
                 {
-                    ALOGI("bt vendor lib: BCM4335 chip detected, needs to check for the lmp version...");
+                    ALOGI("bt vendor lib: %s chip detected, needs to check for the lmp version...", hw_cfg_cb.local_chip_name);
 
                     /* read local revision to check lmp version to differentiate between A0 and B0 revision of BCM4335 */
                     UINT16_TO_STREAM(p, HCI_READ_LOCAL_VERSION_INFORMATION);
@@ -836,10 +837,20 @@ void hw_config_cback(void *p_mem)
                     p_lmp = (uint8_t *) (p_evt_buf + 1) + HCI_EVT_CMD_CMPL_LOCAL_REVISION;
                     STREAM_TO_UINT16(lmp_subversion, p_lmp);
                     ALOGI("bt vendor lib: lmp version : %04x.", lmp_subversion);
-                    if (lmp_subversion == 0x4106)
+                    if (strstr(hw_cfg_cb.local_chip_name, "BCM4335") != NULL)
                     {
-                        /* Found BCM4335B0 revision */
-                        hw_cfg_cb.local_chip_name[7] = 'B';
+                        if (lmp_subversion == 0x4106)
+                        {
+                            /* Found BCM4335B0 revision */
+                            hw_cfg_cb.local_chip_name[7] = 'B';
+                        }
+                    }
+                    if (strstr(hw_cfg_cb.local_chip_name, "BCM43241") != NULL)
+                    {
+                        /* Found BCM43241B3/B4 revision */
+                        /* HCD file still labelled as BCM43241B3, so set accordingly */
+                        if (lmp_subversion == 0x4406)
+                            hw_cfg_cb.local_chip_name[9] = '3';
                     }
                 }
                 /* fall through intentionally */
-- 
1.7.4.1

